### PR TITLE
Improve the interaction of `@Owning` and exceptions

### LINKED ISSUE #6179: Resource Leak Checker: inconsistent behavior for ownership transfer in constructor
Consider:

```java
import org.checkerframework.checker.calledmethods.qual.EnsuresCalledMethods;
import org.checkerframework.checker.mustcall.qual.Owning;
import org.checkerframework.checker.nullness.qual.NonNull;

import java.io.Closeable;
import java.io.IOException;

public class ConstructorOwnership implements Closeable {

    private final @Owning @NonNull Closeable resource;

    public ConstructorOwnership(@Owning @NonNull Closeable resource, int value) {
        if (value == 0) {
            throw new IllegalArgumentException();
        }
        this.resource = resource;
    }

    public ConstructorOwnership(@Owning @NonNull Closeable resource, boolean value) {
        this.resource = resource;
        if (value) {
            throw new IllegalArgumentException();
        }
    }

    @Override
    @EnsuresCalledMethods(value = "resource", methods = {"close"})
    public void close() throws IOException {
        resource.close();
    }

}
```

The resource leak checker reports an error for the first constructor, but not the second:
```
ConstructorOwnership.java:[14,59] error: [required.method.not.called] @MustCall method close may not have been invoked on resource or any of its aliases.
```

I think it should report an error for both constructors or neither. I am not sure which it should be.

### PR DESCRIPTION
**Pull request by Calvin-L**
Merge after https://github.com/codespecs/daikon/pull/524

Fixes #6179.

This PR changes the way that `@Owning` interacts with exceptions in a few subtle (but important!) ways:
 - At a method call, `@Owning` parameters do not satisfy must-call obligations if the method throws an exception.
 - Similarly, within a method body, obligations on `@Owning` parameters only need to be satisfied on normal returns.
 - Within a constructor, assignments to `@Owning` fields only satisfy must-call obligations if the method returns normally.

I regret how large this PR became; I found I had to to do a fair amount of refactoring. To help guide you, I think there are two key changes needing attention:
 1. I reverted one of @kelloggm's changes to treat `@Owning` parameters as if they call the argument's must-call methods. There are three reasons to revert it:
   a. I believe it is unsound. `@Owning` is more like a promise to call those methods later (e.g. if the parameter is assigned to a field). They are not necessarily called before the method exits.
   b. It interacts poorly with the changes in this PR (for reasons I don't entirely have the patience to write up).
   c. It seems safe to revert, since it doesn't change the behavior of any test case except the single one written for it.
 2. The changes in `MustCallConsistencyAnalyzer.updateObligationsForAssignment` are not trivial. While most of the refactoring in that class is straightforward, I had to re-work the logic in `updateObligationsForAssignment` substantially. The old logic seemed to have several latent bugs. Most notably, the (old, now-removed) condition `enclosingCtr.getKind() != ElementKind.CONSTRUCTOR` was always true, and I had to devise a different check to determine if the assignment was happening inside a constructor.

I will also note that I think this PR caught a previously-unknown resource leak in the Checker Framework itself. See the changes I made to [`JavaDiagnosticReader`](https://github.com/typetools/checker-framework/commit/8d38789bb6b6e99da31953fb2d4ecc1c6cd6c7f4#diff-b1f471ce48610691731bb5b8f8d6274423d70f0a65a5cecda5b8c72649395a0f): if the call to `advance()` threw an exception, then the allocated `Reader` would never be closed.

### CODE REVIEW

**Review by kelloggm — CHANGES_REQUESTED** [review]
Overall this is a nice set of changes. I think the refactoring in `MustCallConsistencyAnalyzer` is a clear improvement. Thanks!

My comments are mostly minor, and can probably be addressed mostly through documentation. @msridhar or @Nargeshdb should review as well.

**Comment by Calvin-L** [comment]
CI is failing because of (what I think is) a newly-discovered bug in Daikon:

In [Session.java](https://github.com/codespecs/daikon/blob/aedc5c1794e3fa9dea2a036653babf86ae4671e4/java/daikon/simplify/Session.java#L115), `trace_file` may leak if subsequent lines in the constructor throw `IOException`.

```
./daikon/simplify/Session.java:115: error: [builder:required.method.not.called] @MustCall method close may not have been invoked on trace_file or any of its aliases.
        trace_file = new PrintStream(new FileOutputStream(f));
        ^
  The type of object is: java.io.@org.checkerframework.checker.nullness.qual.MonotonicNonNull PrintStream.
  Reason for going out of scope: possible exceptional exit due to is.read(buf) with exception type java.io.IOException
1 error
```

I'm not quite sure how to resolve this cross-repository check. Do I need to open a PR against Daikon as well?

File: docs/manual/resource-leak-checker.tex
Comment (Calvin-L) [reply docs/manual/resource-leak-checker.tex:230]: Yes, great catch. I had meant to add just such a paragraph. :)

File: checker/tests/resourceleak/OwningEnsuresCalledMethods.java
Comment (Calvin-L) [reply checker/tests/resourceleak/OwningEnsuresCalledMethods.java:2]: Great idea. I've restored this test with an updated comment and set of expected warnings.

**Comment by kelloggm** [comment]
> I'm not quite sure how to resolve this cross-repository check. Do I need to open a PR against Daikon as well?

The CI systems for Daikon and the Checker Framework are linked, so if you open a PR against Daikon with the same branch name in fork owned by the same user account, your changes in both repositories will be used when running CI on each.

You could also ask @mernst (or another Daikon maintainer) to look into or fix the newly-discovered Daikon bug, if you're not sure what the correct patch is.

**Review by kelloggm — APPROVED** [review]

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:1186]: I would be a bit surprised is there isn't a more efficient way to do this.  If needed, I would be in favor of storing the `cfg` passed to `analyze()` in an instance field so we can check this type of condition more efficiently.  See also #5985

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:1229]: Is this shifting around some extant logic, or handling a new case?  If it's the latter, what is the motivation?

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:1332]: Is there a reason we need an `enum` here and this can't just be a boolean parameter to `removeObligationsContainingVar()`?

**Review by msridhar — COMMENTED** [review]
This is a really great change.  I'm not through the whole PR yet but thought I'd submit my current comments; hope to get through the rest by the end of the day

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java
Comment (Calvin-L) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:1332]: No reason other than [personal stylistic preference](https://apaloski.github.io/coding/style/enum-vs-boolean).

At a glance, (I think) this code is more difficult to read:
```java
          removeObligationsContainingVar(
              ...
              false, // no shred of a clue what this argument does :(
              ...);
```

Comment (msridhar) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:1332]: Interesting, a two-valued enum to kind of simulate a named parameter!  I'm fine with keeping it here 🙂

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java
Comment (Calvin-L) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:1229]: This is handling a new case, and I'm glad you called it out because I should have mentioned it in the PR description.

Another big change here is that `ResourceAlias` can contain non-local-variable expressions. This is somewhat dangerous, so the only place that happens is right here, where final member fields are added to obligations' resource-alias sets.

The motivation is to support this test case in `OwnershipWithExceptions.java`:
```java
    FinalOwnedField(int ignored) throws IOException {
      // Same as the 0-argument constructor, but handled correctly (algorithm 1).
      resource = alloc();
      try {
        if (arbitraryChoice()) {
          throw new IOException();
        }
      } catch (Exception e) {
        resource.close();
        throw e;
      }
    }
```

Without this change, the call to `resource.close()` does not satisfy the RLC that the value returned by `alloc()` has been closed.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java
Comment (Calvin-L) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java]: Is [3dff5b6](https://github.com/typetools/checker-framework/pull/6241/commits/3dff5b69fd2d159c6c3394dde4d5d6f07ecb7038) what you had in mind?

Comment (msridhar) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java]: Yes, this exactly, thanks!

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:1989]: We now redundantly compute these obligation updates once per successor block.  On the other hand, given that `obligations` is mutable, it's not obvious how to cache these results.  And, I don't even know if these computations are all that expensive.  Maybe we should add a comment here indicating we are at least aware of the redundancy and that optimization may be worthwhile in the future if we find this is a bottleneck?

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:2368]: Does this work?
```suggestion
                .getAnnotatedType(alias.element)
```

File: checker/tests/resourceleak/Generics.java
Comment (msridhar) [inline checker/tests/resourceleak/Generics.java:1]: How is this test case related to this PR?

File: checker/tests/resourceleak/OwningEnsuresCalledMethods.java
Comment (msridhar) [inline checker/tests/resourceleak/OwningEnsuresCalledMethods.java:12]: 👍

**Review by msridhar — COMMENTED** [review]
A few more minor comments but otherwise this looks great to me!

File: checker/tests/resourceleak/Generics.java
Comment (Calvin-L) [reply checker/tests/resourceleak/Generics.java:1]: At one point during development I thought I had a bug related to generics. It turned out to be nothing. This test had some sketchy code to help me figure out what was going on, and now serves no purpose. I'll remove it.

**Review by msridhar — APPROVED** [review]
Thanks again for this contribution!

**Comment by kelloggm** [comment]
Both should be merged at the same time, so we need to wait for the Daikon PR to get reviewed.

**Pull request merged**
